server/src/main/java/io/crate/expression/scalar/conditional/CaseFunction.java

```diff
@@ -0,0 +1,109 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar.conditional;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+import java.util.List;
+
+import io.crate.data.Input;
+import io.crate.expression.FunctionExpression;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.expression.symbol.Literal;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public class CaseFunction extends Scalar<Object, Object> {
+
+    public static final String NAME = "case";
+
+    public static void register(ScalarFunctionModule module) {
+        // case([boolean] conditions, [E] results) -> E
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.BOOLEAN_ARRAY.getTypeSignature(),
+                parseTypeSignature("array(E)"),
+                parseTypeSignature("E")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            CaseFunction::new
+        );
+        // case([boolean] conditions, [E] results, E default) -> E
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.BOOLEAN_ARRAY.getTypeSignature(),
+                parseTypeSignature("array(E)"),
+                parseTypeSignature("E"),
+                parseTypeSignature("E")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            CaseFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    private CaseFunction(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>[] args) {
+        List<Boolean> conditions = (List<Boolean>) args[0].value();
+        for (int i = 0; i < conditions.size(); i++) {
+            var condition = conditions.get(i);
+            if (condition != null && condition) {
+                if (args[1] instanceof FunctionExpression<?,?> fe) {
+                    return fe.getArguments()[i].value();
+                } else if (args[1] instanceof Literal<Object> lt) {
+                    return ((List) lt.value()).get(i);
```

Comment:
@matriv We could have a generic value(int position) method over Input, which would be then implemented by Literal and FunctionExpression etc.

Reply:
sounds good